pulp_container/app/tasks/sync_stages.py

```diff
@@ -126,20 +160,54 @@ async def run(self):
                     list_dc = self.create_tagged_manifest_list(
                         tag_name, saved_artifact, content_data
                     )
-                    await self.put(list_dc)
-                    tag_dc.extra_data["tagged_manifest_dc"] = list_dc
+                    at_least_one_manifest = False
                     for manifest_data in content_data.get("manifests"):
                         man_dc = self.create_manifest(list_dc, manifest_data)
+                        if signature_source is not None:
+                            man_sig_dcs = await self.create_signatures(man_dc, signature_source)
+                            if self.signed_only and not man_sig_dcs:
+                                log.info(
+                                    _(
+                                        "The unsigned image {img_digest} which is a part of the "
+                                        "manifest list {ml_digest} (tagged as `{tag}`) is "
+                                        "skipped due to a requirement to sync signed content "
+                                        "only.".format(
+                                            img_digest=man_dc.d_content.digest,
+                                            ml_digest=list_dc.d_content.digest,
+                                            tag=tag_name,
+                                        )
+                                    )
+                                )
+                                # do not pass down the pipeline unsigned manifests
+                                continue
+                            signature_dcs.extend(man_sig_dcs)
+                        at_least_one_manifest = True
                         future_manifests.append(man_dc)
                         man_dcs[man_dc.content.digest] = man_dc
                         await self.put(man_dc)
+
+                    if not at_least_one_manifest:
+                        # no manifests for this manifest list and tag were signed, do not pass
+                        # this manifest list and tag down the pipeline
+                        continue
+
+                    tag_dc.extra_data["tagged_manifest_dc"] = list_dc
```

Comment:
tag_dc is being appended to the tag_dcs list  to go through and resolve its `tagged_manifest_dc`regardless of whether the list_dc( or manifest_dc) was sent through the pipeline or not.

if i read correctly, this line will fail because it expects to have `tagged_manifest_dc` among the extra_data https://github.com/pulp/pulp_container/pull/436/files#diff-b702cbd3ebdff19bf8e4dfc14688aa3fb36e1a47ef6cf84d24e0f75b61247d36R224 which is not there.
we should change that line to `tagged_manifest_dc = tag_dc.extra_data.get("tagged_manifest_dc")`

Reply:
I do not think tag_dc is being appended if manifest list or manifest do not go through, at least it was my intention not to sync that tag at all.

Line-192 `continue` makes sure that manifest list and its tag are not passed further (including the appending to tag_dcs).
Line-205 `continue` makes sure that that a tag of a manifest is not passed further (including the appending to tag_dcs).
I expect the `tagged_manifest_dc` to always have a manifest in extra_data.

Let me know if I misunderstood your concern.

Reply:
no, my concern was understood correctly. it was difficult to read nested loops, `continue` on lines 192 and 205 match the out loop so all is :+1: